### DISCUSSION TITLE
Early Return은 과연 좋은가? | 아카이브-로그

### DISCUSSION THREAD

**Discussion by giscus** [jongwooo/blog#361] [Comment]
# Early Return은 과연 좋은가? | 아카이브-로그

코드를 살펴 보던 중, 아래와 같이 불필요하게 중첩된 조건문을 발견하였습니다. 위 코드를 아래와 같이 Early Return하는 코드로 변경하였습니다. Early Return은 함수에서 조건문을 만족할 때 일찍 반환하는 것을 의미합니다. 위의 변경사항을 커밋하여 풀 리퀘스트를 생성하니, 리뷰어께서 아래와 같은 의견을 남겨주셨습니다. I think an e…

https://thearchivelog.dev/article/are-early-returns-any-good/

**Comment by sunglim** [comment]
오해할것같은데, 대부분의 경우 early return이 좋습니다 
저 리뷰어가 코멘트가 남긴건
잘 돌아가고 있는 코드에  refactoring을 하니  한마디 남긴것뿐이라고 생각합니다

**Reply by sunglim** [reply]
특히 early return이 복잡도를 높이는건 절대아닙니다 if 2depths가 더 복잡도를 높이죠

**Reply by jongwooo** [reply]
@sunglim 안녕하세요! 댓글 주셔서 감사합니다.
해당 글은 불필요하게 중첩된 조건문을 수정하는 과정에서, 리뷰어께서 남겨주신 코멘트를 통해서 고민한 점을 정리한 것입니다.

early return이 단점이 있으니 if 2depths를 사용하라는 의도가 아니라는 점 알아주셨으면 합니다.
early return도 결국 함수 라인 수를 높일 수도 있으니 은탄환이 될 수 없죠.
메서드 추출과 같은 방법을 사용하면 더 좋을 것 같네요:)

**Reply by xoxwgys56** [reply]
@jongwooo 좋은 글 감사합니다. 덕분에 early-return에 대해서 좀 더 생각해볼 수 있는 기회를 얻게 되었습니다.  
말씀하신 메서드 추출도 좋은 해결 방법이라고 생각합니다.

상황에 따라 필요한지를 선택하는 것이 개발자에게 필요한 요소라고 생각합니다.  
혹시 early-return 상대적으로 유리한 경우에 대해 몇가지 예시를 공유해주실 수 있을까요?  
글을 이해하는데 더 도움이 될 것 같아 요청드립니다.

다시 한번 좋은 글에 대해 감사드립니다 :)

**Reply by jongwooo** [reply]
@xoxwgys56 안녕하세요! 댓글 주셔서 감사합니다.
Early Return은 [Arrow Anti Pattern](https://wiki.c2.com/?ArrowAntiPattern)을 해결하는 대표적인 방법이죠. 하지만 이를 안티 패턴이고 문제라고 생각하는 것이 먼저인 것 같아요. '이렇게 작성해도 상관 없다', '문제가 아니다'라고 한다면 그냥 넘어갈 수도 있는 코드니까요. 일단 돌아가면 그대로 두라는 말처럼 말이죠. 개인적으로 '상대적으로 유리한 경우' 보다는, 팀이나 개인이 코드를 어떻게 작성하겠다고 한 약속(개인은 다짐이겠죠)에 따라 사용이 달라진다고 생각합니다.

**Comment by rbgusdlza** [comment]
때때로 Early Return은 함수의 복잡도를 증가시키고 코드를 더 복잡하게 만들 수 있습니다. 

-> 이런 경우가 있을까요? 오히려 early return을 사용하면 depth도 줄고 if 블럭을 따로 메서드 추출할 수 있어서 복잡성이 줄고 가독성도 증가할텐데요

**Reply by jongwooo** [reply]
@rbgusdlza 안녕하세요! 댓글 주셔서 감사합니다.

말씀주신 부분에서 "if 블럭을 따로 메서드 추출할 수 있다"는 bouncer pattern을 통해 빠른 예외 처리를 할 수 있다는 점에서 동의합니다. 의견에 깊이 공감합니다만, 제가 우려하는 부분은 예외 처리가 아닌 특정 값 자체를 리턴해야 하는 경우에는 여러 개의 return 문을 작성해야 한다는 것입니다. 이는 하나의 출구(return)가 아니기 때문에 어느 부분에서 값이 반환되는지 코드를 일일히 살펴보아야 확인할 수 있을 것이라고 생각합니다.
